### PR TITLE
#A04 - As Admin, I would like to delete existing Hackathon events

### PR DESCRIPTION
**Pull request by hebs87**
## User Story ID:
- #A04 (As Admin, I would like to delete existing Hackathon events)
- Feature

## Description of PR:
All changes made are outlined below:
  - Added status field to Hackathon model to enable soft deleting Hackathon event
  - Added delete_hackathon view to set hackathon status to 'deleted' to remove it from the frontend list but keep it in the DB:
    - If user is not admin user, they are redirected to the hackathons list page
  - Added new URL for the view with the 'hackathon_id' arg to enable the correct event to be deleted in the view
  - Amended HackathonListView to pass in the following context to the hackathon_list.html template:
    - hackathons - filter out 'deleted' hackathons and order by newest first
    - today - needed to compare dates to conditionally display message in delete alert for hackathons in progress (needed because using built in template 'now' date didn't work correctly for the comparison)
  - Amended hackathon_list.html template to use new 'hackathons' context to render only non-deleted hackathons
  - Amended hackathon_card.html with the following:
    - Added 'DELETE' button (only for admin users) to each rendered hackathon card, which triggers a confirmation modal when clicked
    - Confirmation delete modal to confirm delete request with the following logic:
      - For past or future hackathons, the message asks user if they are sure they want to delete the relevant hackathon
      - For live hackathons, the message warns user that the hackathon is live and asks them if they want to delete it
      - If user clicks 'NO', the modal is closed and the event remains in the list and not deleted
      - If user clicks 'YES', the hackathon's status is set to 'deleted' and is no longer rendered in the frontend

## Tests:
Manually tested all implemented functionality, which is all working as expected.
Validated all code in relevant validators and all passed - HTML validator threw errors for Django template tags, but no other errors.

## Know bugs/errors:
No known bugs or errors after testing.

## Screenshots:
image

image

image

## Does this introduce a breaking change
- [X] Yes
- [] No

### CODE REVIEW

File: hackathon/models.py
Comment (stefdworschak) [reply hackathon/models.py:46]: Also, this should use a list of choices. You can have a look at @auxfuse's PR and just adjust to the same definition: https://github.com/Code-Institute-Community/ci-hackathon-app/pull/101

File: hackathon/templates/hackathon/includes/hackathon_card.html
Comment (stefdworschak) [inline hackathon/templates/hackathon/includes/hackathon_card.html:7]: What does that mean? Is this for testing purposes? If it is please remove it.

File: hackathon/templates/hackathon/includes/hackathon_card.html
Comment (stefdworschak) [inline hackathon/templates/hackathon/includes/hackathon_card.html:29]: Please change the id to something more meaningful or remove it if not needed.

**Review by stefdworschak — CHANGES_REQUESTED** [review]
@hebs87 looks good in general! Have a look at my comments and let me know once I can have another look at it. :+1:

File: hackathon/templates/hackathon/includes/hackathon_card.html
Comment (hebs87) [reply hackathon/templates/hackathon/includes/hackathon_card.html:7]: Yep, it was for testing purposes to figure out how I could make the date logic work in the modal. I just forgot to take it out! I've removed it now and it should be visible when I push the changes to the existing PR.

File: hackathon/templates/hackathon/includes/hackathon_card.html
Comment (hebs87) [reply hackathon/templates/hackathon/includes/hackathon_card.html:29]: This was the default id from the Bootstrap modal template, I missed it. I've removed it now and the change will be visible when I push it to the existing PR 👍

File: hackathon/models.py
Comment (hebs87) [reply hackathon/models.py:46]: Cool, so from the PR that you've tagged, it looks like a status field has already been added to the Hackathon model in that PR, with the relevant choices. With this being the case, I'll remove my change from this file so that it doesn't cause any merge conflicts. Then, once both PRs have been merged in to the master branch, it should work fine - the logic in the delete_hackathon view is already correct to work with the relevant status choices - sets the status to 'deleted', which is one of the choices. If that's cool with you, just let me know and I'll go ahead and remove my changes to this file and push it.

Comment (hebs87) [reply hackathon/models.py:46]: Alternatively, I'm happy to wait until the other PR has been merged in. Once done, I can then pull that in and test that my code integrates correctly with it before pushing any of my final changes to this PR. I'm cool with either option, just let me know please 👍

**Comment by stefdworschak** [comment]
@hebs87 I merged @auxfuse's pull request, could you merge his changes into your branch and update the PR? Thanks

File: hackathon/models.py
Comment (hebs87) [reply hackathon/models.py]: @stefdworschak This is now integrated with the latest PR - tested and working.

**Review by stefdworschak — APPROVED** [review]
LGTM

**Pull request merged**
